File: platform/android/lynx_android/src/main/java/com/lynx/tasm/gesture/handler/FlingGestureHandler.java
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.gesture.handler;

import android.view.MotionEvent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.lynx.react.bridge.ReadableMap;
import com.lynx.tasm.behavior.LynxContext;
import com.lynx.tasm.event.LynxTouchEvent;
import com.lynx.tasm.gesture.GestureArenaMember;
import com.lynx.tasm.gesture.detector.GestureDetector;
import java.util.HashMap;

/**
 * The FlingGestureHandler class is a concrete implementation of the BaseGestureHandler class
 * for handling fling gestures. It handles fling gesture for component, triggered when the finger is
 * lifted and speed is not zero.
 */
public class FlingGestureHandler extends BaseGestureHandler {
  // is onBegin invoked or not
  private boolean mIsInvokedBegin = false;
  // is onStart invoked or not
  private boolean mIsInvokedStart = false;
  // is onEnd invoked or not
  private boolean mIsInvokedEnd = false;

  private final HashMap<String, Object> mEventParams;

  /**
   * Constructs a FlingGestureHandler object with the specified properties.
   *
   * @param sign                the sign indicating the direction of the gesture
   * @param lynxContext         the LynxContext associated with the gesture handler
   * @param gestureDetector     the GestureDetector associated with the gesture handler
   * @param gestureArenaMember  the GestureArenaMember associated with the gesture handler
   */
  public FlingGestureHandler(int sign, LynxContext lynxContext,
      @NonNull GestureDetector gestureDetector, GestureArenaMember gestureArenaMember) {
    super(sign, lynxContext, gestureDetector, gestureArenaMember);
    handleConfigMap(gestureDetector.getConfigMap());
    mEventParams = new HashMap<>();
  }

  @Override
  protected void handleConfigMap(@Nullable ReadableMap config) {
    if (config == null) {
      return;
    }
  }

  @Override
  protected void onHandle(@Nullable MotionEvent event, @Nullable LynxTouchEvent lynxTouchEvent,
      float flingDeltaX, float flingDeltaY) {
    if (mGestureArenaMember != null) {
      mGestureArenaMember.onInvalidate();
    }

    if (event != null
        && (event.getActionMasked() == MotionEvent.ACTION_DOWN
            || event.getActionMasked() == MotionEvent.ACTION_MOVE)) {
      // If the event is not empty, it means the finger on the screen, no need to handle fling
      // gesture
      ignore();
      return;
    }

    if (event != null && event.getActionMasked() == MotionEvent.ACTION_UP) {
      begin();
      onBegin(0, 0, null);
      return;
    }

    if (mStatus >= GestureConstants.LYNX_STATE_FAIL && mStatus <= GestureConstants.LYNX_STATE_END) {
      onEnd(0, 0, null);
      return;
    }

    if (flingDeltaX == Float.MIN_VALUE && flingDeltaY == Float.MIN_VALUE) {
      fail();
      onEnd(0, 0, null);
      return;
    }

    if (mStatus == GestureConstants.LYNX_STATE_INIT
        || mStatus == GestureConstants.LYNX_STATE_UNDETERMINED) {
      begin();
      activate();
      onBegin(0, 0, null);
      onStart(0, 0, null);
      return;
    }

    onUpdate(flingDeltaX, flingDeltaY, null);
  }

  protected HashMap<String, Object> getEventParamsInActive(float deltaX, float deltaY) {
    mEventParams.put("scrollX", px2dip(mGestureArenaMember.getMemberScrollX()));
    mEventParams.put("scrollY", px2dip(mGestureArenaMember.getMemberScrollY()));
    mEventParams.put("deltaX", px2dip(deltaX));
    mEventParams.put("deltaY", px2dip(deltaY));
    mEventParams.put("isAtStart", mGestureArenaMember.isAtBorder(true));
    mEventParams.put("isAtEnd", mGestureArenaMember.isAtBorder(false));
    return mEventParams;
  }

  @Override
  public void fail() {
    super.fail();
    onEnd(0, 0, null);
  }

  @Override
  public void end() {
    super.end();
    onEnd(0, 0, null);
  }

  @Override
  public void reset() {
    super.reset();
    mIsInvokedBegin = false;
    mIsInvokedStart = false;
    mIsInvokedEnd = false;
  }

  @Override
  protected void onBegin(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnBeginEnable() || mIsInvokedBegin) {
      return;
    }
    mIsInvokedBegin = true;
    sendGestureEvent(GestureConstants.ON_BEGIN, getEventParamsInActive(x, y));
  }

  @Override
  protected void onUpdate(float deltaX, float deltaY, @Nullable LynxTouchEvent event) {
    if (!isOnUpdateEnable()) {
      return;
    }
    sendGestureEvent(GestureConstants.ON_UPDATE, getEventParamsInActive(deltaX, deltaY));
  }

  @Override
  protected void onStart(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnStartEnable() || mIsInvokedStart || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedStart = true;
    sendGestureEvent(GestureConstants.ON_START, getEventParamsInActive(x, y));
  }

  @Override
  protected void onEnd(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnEndEnable() || mIsInvokedEnd || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedEnd = true;
    sendGestureEvent(GestureConstants.ON_END, getEventParamsInActive(x, y));
  }
}

File: clay/ui/gesture_handler/handler/fling_gesture_handler.cc
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "clay/ui/gesture_handler/handler/fling_gesture_handler.h"

#include <limits>
#include <utility>

#include "clay/public/value.h"
#include "clay/ui/event/gesture_event.h"
#include "clay/ui/gesture_handler/gesture_arena_member.h"

namespace clay {

FlingGestureHandler::FlingGestureHandler(
    int sign, PageView* page_view,
    std::shared_ptr<GestureDetector> gesture_detector,
    fml::WeakPtr<GestureArenaMember> gesture_arena_member)
    : BaseGestureHandler(sign, page_view, gesture_detector,
                         gesture_arena_member),
      is_invoked_begin_(false),
      is_invoked_start_(false),
      is_invoked_end_(false) {
  HandleConfigMap(gesture_detector->config_map());
}

void FlingGestureHandler::HandleConfigMap(const Value& config) {}

void FlingGestureHandler::OnHandle(
    const PointerEvent* pointer_event, float fling_delta_x, float fling_delta_y,
    bool handle_by_simultaneous,
    const std::shared_ptr<GestureExtraBundle>& extra_bundle) {
  if (pointer_event != nullptr &&
      (pointer_event->type == PointerEvent::EventType::kDownEvent ||
       pointer_event->type == PointerEvent::EventType::kMoveEvent)) {
    // If the event is not empty, it means the finger on the screen, no need to
    // handle fling gesture
    Ignore();
    return;
  }
  if (pointer_event != nullptr &&
      pointer_event->type == PointerEvent::EventType::kUpEvent) {
    Begin();
    OnBegin(0, 0, nullptr);
    return;
  }

  if (status_ >= GestureConstants::LYNX_STATE_FAIL &&
      status_ <= GestureConstants::LYNX_STATE_END) {
    OnEnd(0, 0, nullptr);
    return;
  }

  if (fling_delta_x == std::numeric_limits<float>::min() &&
      fling_delta_y == std::numeric_limits<float>::min()) {
    Fail();
    OnEnd(0, 0, nullptr);
    return;
  }

  if (status_ == GestureConstants::LYNX_STATE_INIT ||
      status_ == GestureConstants::LYNX_STATE_UNDETERMINED) {
    Begin();
    Activate();
    OnBegin(0, 0, nullptr);
    OnStart(0, 0, nullptr);
    return;
  }

  OnUpdate(fling_delta_x, fling_delta_y, nullptr, extra_bundle);
}

void FlingGestureHandler::Fail() {
  if (status_ != GestureConstants::LYNX_STATE_FAIL) {
    status_ = GestureConstants::LYNX_STATE_FAIL;
    OnEnd(0, 0, nullptr);
  }
}

void FlingGestureHandler::End() {
  if (status_ != GestureConstants::LYNX_STATE_END) {
    status_ = GestureConstants::LYNX_STATE_END;
    OnEnd(0, 0, nullptr);
  }
}

void FlingGestureHandler::Reset() {
  BaseGestureHandler::Reset();
  is_invoked_begin_ = false;
  is_invoked_start_ = false;
  is_invoked_end_ = false;
}

Value FlingGestureHandler::GenerateAdditionalEventParams(float delta_x,
                                                         float delta_y) {
  Value::Map res;
  res["scrollX"] = Value(gesture_arena_member_->ScrollX());
  res["scrollY"] = Value(gesture_arena_member_->ScrollY());
  res["deltaX"] = Value(delta_x);
  res["deltaY"] = Value(delta_y);
  res["isAtStart"] = Value(gesture_arena_member_->IsAtBorder(true));
  res["isAtEnd"] = Value(gesture_arena_member_->IsAtBorder(false));
  return Value(std::move(res));
}

void FlingGestureHandler::OnBegin(float x, float y,
                                  const PointerEvent* pointer_event) {
  if (!IsOnBeginEnable() || is_invoked_begin_) {
    return;
  }
  is_invoked_begin_ = true;
  Value addition_params = GenerateAdditionalEventParams(x, y);
  SendGestureEvent(GestureConstants::ON_BEGIN, pointer_event, addition_params);
}

void FlingGestureHandler::OnUpdate(
    float delta_x, float delta_y, const PointerEvent* pointer_event,
    const std::shared_ptr<GestureExtraBundle>& extra_bundle) {
  if (!IsOnUpdateEnable()) {
    return;
  }
  Value addition_params = GenerateAdditionalEventParams(delta_x, delta_y);
  SendGestureEvent(GestureConstants::ON_UPDATE, pointer_event, addition_params);
}

void FlingGestureHandler::OnStart(float x, float y,
                                  const PointerEvent* pointer_event) {
  if (!IsOnStartEnable() || is_invoked_start_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_start_ = true;
  Value addition_params = GenerateAdditionalEventParams(x, y);
  SendGestureEvent(GestureConstants::ON_START, pointer_event, addition_params);
}

void FlingGestureHandler::OnEnd(float x, float y,
                                const PointerEvent* pointer_event) {
  if (!IsOnEndEnable() || is_invoked_end_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_end_ = true;
  Value addition_params = GenerateAdditionalEventParams(x, y);
  SendGestureEvent(GestureConstants::ON_END, pointer_event, addition_params);
}

}  // namespace clay
